decode_arm64.s

```diff
@@ -182,34 +142,45 @@ readMatchlenDone:
 	// but v3 doesn't support dictionary files.
 	BLT dictNotSupported
 
-	// Copy a regular match.
-	// Since len+minMatch is at least four, we can do a 4× unrolled
-	// byte copy loop. Using MOVW instead of four byte loads is faster,
-	// but to remain portable we'd have to align match first, which is
-	// too expensive. By alternating loads and stores, we also handle
-	// the case offset < 4.
-copyMatch4:
-	SUBS    $4, len
+	// Copy doublewords if both len and offset are at least eight.
+	// A 16-at-a-time loop doesn't provide a further speedup.
+	CMP  $8, len
+	CCMP HS, offset, $8, $0
+	BLO  copyMatchLoop1
+
+	AND    $7, len, lenRem
+	SUB    $8, len
+copyMatchLoop8:
+	SUBS   $8, len
+	MOVD.P 8(match), tmp1
+	MOVD.P tmp1, 8(dst)
+	BPL    copyMatchLoop8
+
+	ADD  lenRem, match
+	ADD  lenRem, dst
+	MOVD -8(match), tmp2
+	MOVD tmp2, -8(dst)
+	B    copyMatchDone
+
+	// 4× unrolled byte copy loop for the overlapping case.
+copyMatchLoop4:
+	SUB     $4, len
 	MOVBU.P 4(match), tmp1
 	MOVB.P  tmp1, 4(dst)
 	MOVBU   -3(match), tmp2
 	MOVB    tmp2, -3(dst)
 	MOVBU   -2(match), tmp3
 	MOVB    tmp3, -2(dst)
-	MOVBU   -1(match), tmp1
-	MOVB    tmp1, -1(dst)
-	BPL     copyMatch4
-
-	// Restore len, which is now negative.
-	ADDS  $4, len
-	BEQ   copyMatchDone
+	MOVBU   -1(match), tmp4
+	MOVB    tmp4, -1(dst)
+	CBNZ   len, copyMatchLoop4
```

Comment:
I put one space too few here, maybe correct that?

Reply:
done.